### PR TITLE
Gallery View & Create Folder uses GraphQL (fixes #316)

### LINKED ISSUE #316: Gallery View & Create Folder uses GraphQL
As a developer, I want to use GraphQL as the API for the asset section, so that I have a consistent content API.

## Acceptance criteria:

 * Gallery view fetches files via GraphQL
 * Folder creation occurs via GraphQL

## Exclusions

 * `File::canView()` and `Folder::canCreate()` permissions are not checked (see https://github.com/silverstripe/silverstripe-graphql/issues/5)
 * ~~Pagination in gallery view~~ (included this now)

## Tasks

 * [x] Rebase on master
 * [x] Pagination in list view (see [issue](https://github.com/silverstripe/silverstripe-asset-admin/issues/278))
 * [x] Pagination in gallery view (see [issue](https://github.com/silverstripe/silverstripe-asset-admin/issues/278))
 * [x] Sort on files (dropdown and table sort)
 * [x] Respect sort when adding new folder after mutation
 * [x] Fix folder navigation (currently defaults to `undefined`) - might be fixed by [rebase](https://github.com/silverstripe/silverstripe-asset-admin/issues/316)
 * [x] Fix persistence of uploaded files in current view
 * [x] Only refetch() once after all uploads complete (not every one)
 * [x] Don't show "no items" when loading data (e.g. during pagination)
 * [x] Merge custom sort for GraphQL (https://github.com/silverstripe/silverstripe-graphql/pull/24)
 * [x] Merge case accessor patch (https://github.com/silverstripe/silverstripe-graphql/pull/31)
 * [x] Test with "select file" in UploadField
 * [x] Fix existing unit tests
 * [x] Write at least one GraphQL related unit test (e.g. GraphQL query triggered on folder delete)
 * [x] Fix behat tests
 * [x] Cross browser tests
 * [ ] Pagination is broken when uploading first file
 * [ ] GraphQL resolver allows duplicating folder names
 * [ ] PHPDoc for GraphQL classes in asset-admin

## Notes

 * (Was: Rebase and merge GraphQL work)
 * See https://github.com/silverstripe/silverstripe-framework/issues/6245. Applies to both `framework` and `asset-admin` repos.

### PR DESCRIPTION
**Pull request by chillu**
This is the first introduction of GraphQL and Apollo as core dependencies (see [RFC](https://github.com/silverstripe/silverstripe-framework/issues/6245)). It also adds a dependency on https://github.com/silverstripe/silverstripe-graphql.

Pushing this up for peer review at the moment, see outstanding tasks in https://github.com/silverstripe/silverstripe-asset-admin/issues/316

Relies on framework PR: https://github.com/silverstripe/silverstripe-framework/pull/6420

### CODE REVIEW

**Comment by tractorcow** [comment]
I've removed some dead code from AssetAdmin.php, and I've updated the upload behaviour so that mid-uploaded files no longer temporarily disappear from the screen. :)

File: code/Controller/AssetAdmin.php [hunk 1]
```diff
@@ -61,12 +61,7 @@ class AssetAdmin extends LeftAndMain implements PermissionProvider
         // Pass all URLs to the index, for React to unpack
         'show/$FolderID/edit/$FileID' => 'index',
         // API access points with structured data
-        'POST api/createFolder' => 'apiCreateFolder',
```
Comment (chillu) [inline code/Controller/AssetAdmin.php]: Ah yep, thanks for removing the dead code. I had actually removed it on the experiments/graphql branch, but missed it when "rebasing" by manually applying the changes on master for this new pulls/4.0/graphql branch

File: code/GraphQL/CreateFileMutationCreator.php [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+                }
+            ],
+        ];
+    }
+
+    public function resolve($object, array $args, $context, $info)
+    {
+        $parent = Versioned::get_one_by_stage(Folder::class, 'Stage', $args['file']['parentId']);
+
+        if(!$parent) {
```
Comment (tractorcow) [inline code/GraphQL/CreateFileMutationCreator.php]: What if we are creating a file in the root folder?

File: code/GraphQL/CreateFileMutationCreator.php [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+        return [
+            'file' => [
+                'type' => function() {
+                    return $this->manager->getType('FileInput');
+                }
+            ],
+        ];
+    }
+
+    public function resolve($object, array $args, $context, $info)
```
Comment (tractorcow) [inline code/GraphQL/CreateFileMutationCreator.php]: No PHPDoc on this or on any ancestor implementor.

File: code/GraphQL/CreateFileMutationCreator.php [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+                $args['parentId']
+            ));
+        }
+
+        $file = File::create();
+        foreach($args['file'] as $name => $val) {
+            $this->accessor->setValue($file, $name, $val);
+        }
+
+//        if(!$file->canCreate()) {
```
Comment (tractorcow) [inline code/GraphQL/CreateFileMutationCreator.php]: You can use `File::singleton()->canCreate(null, ['Parent' => $parent]);` if you want to shift this to before the `File::create()`.

**Comment by tractorcow** [comment]
I've also raised https://github.com/silverstripe/silverstripe-graphql/pull/28 with some feedback / cleanup related to this PR.

**Comment by tractorcow** [comment]
I have some PHP fixes to the above issues to push up, but am just waiting for your work to get merged in first @chillu .

File: code/GraphQL/CreateFileMutationCreator.php (hunk 4 above)
Comment (chillu) [inline code/GraphQL/CreateFileMutationCreator.php]: Yep, that's cleaner, done now.

File: code/GraphQL/CreateFileMutationCreator.php (hunk 2 above)
Comment (chillu) [inline code/GraphQL/CreateFileMutationCreator.php]: Good point - the `FileInputTypeCreator` requires a `parentId` field, so we can rely on it here.

**Comment by chillu** [comment]
@tractorcow Merged https://github.com/silverstripe/silverstripe-graphql/pull/28 and responded to your PR feedback.

**Comment by tractorcow** [comment]
Reased

File: code/GraphQL/ReadFileQueryCreator.php [hunk 5]
```diff
@@ -57,6 +57,19 @@ public function connection()
 
     public function resolveConnection($object, array $args, $context, $info)
     {
+        // Permission checks
+        $parent = Folder::singleton();
+        if (isset($args['parentId'])) {
+            $parent = Folder::get()->byID($args['parentId']);
```
Comment (tractorcow) [inline code/GraphQL/ReadFileQueryCreator.php]: Need error handling if parentId isn't a folder.

File: client/src/containers/AssetAdmin/AssetAdmin.js [hunk 6]
```diff
@@ -292,6 +292,7 @@ class AssetAdmin extends SilverStripeComponent {
         },
       ],
     }).then(() => {
+      console.log(this.props.actions.gallery.deselectFiles);
```
Comment (tractorcow) [inline client/src/containers/AssetAdmin/AssetAdmin.js]: Remove debugging code. :)

**Comment by tractorcow** [comment]
I've added thumbnail, width, height, and smallThumbnail to the graphql model; Now we aren't getting thumbnails displaying with full sized images.

**Comment by tractorcow** [comment]
Ok, FINALLY we are green.

@chillu please review my recent changes. They are mostly just fixes to get the CI working. If you approve please merge.

**Pull request merged**